src/embind/embind.js

```diff
@@ -729,39 +729,79 @@ var LibraryEmbind = {
     '$readLatin1String', '$registerType',
     '$simpleReadValueFromPointer'],
   _embind_register_std_wstring: function(rawType, charSize, name) {
-    // nb. do not cache HEAPU16 and HEAPU32, they may be destroyed by emscripten_resize_heap().
     name = readLatin1String(name);
-    var getHeap, shift;
+    var decodeString, encodeString, getHeap, lengthBytesUTF, shift;
     if (charSize === 2) {
+        decodeString = UTF16ToString;
+        encodeString = stringToUTF16;
+        lengthBytesUTF = lengthBytesUTF16;
         getHeap = function() { return HEAPU16; };
         shift = 1;
     } else if (charSize === 4) {
+        decodeString = UTF32ToString;
+        encodeString = stringToUTF32;
+        lengthBytesUTF = lengthBytesUTF32
         getHeap = function() { return HEAPU32; };
         shift = 2;
     }
     registerType(rawType, {
         name: name,
         'fromWireType': function(value) {
-            var HEAP = getHeap();
             var length = HEAPU32[value >> 2];
-            var a = new Array(length);
-            var start = (value + 4) >> shift;
-            for (var i = 0; i < length; ++i) {
-                a[i] = String.fromCharCode(HEAP[start + i]);
+            var HEAP = getHeap();
+            var str;
```

Comment:
I'm relatively ok with this duplication, but would appreciate a comment inline that this is copied from register_std_string's registerType